File: packages/swift/Sources/PylonSync/LocalStore.swift
import Foundation
import PylonClient

public typealias Row = [String: JSONValue]

/// In-memory replica of server state. Mirrors `LocalStore` from
/// `packages/sync/src/index.ts` — same tombstone semantics, same
/// authoritative-id handling on insert/update, same notify model.
///
/// Thread-safe via an internal lock. Listeners are invoked on the caller
/// thread that triggered the change — wrap in `Task { @MainActor in ... }`
/// when bridging to UI frameworks if you need main-thread guarantees.
public final class LocalStore: @unchecked Sendable {
    private let lock = NSLock()
    private var tables: [String: [String: Row]] = [:]

    /// Tombstones: `(entity, row_id) -> deletedAt seq`. A row whose id is
    /// here is considered deleted; any insert/update with `seq < tombSeq`
    /// is dropped so a delayed replay can't resurrect it.
    private var tombstones: [String: [String: Int64]] = [:]
    /// Rows deleted optimistically and not yet confirmed by the server.
    /// Blocks an insert/update for the id until the server's own delete
    /// event lands (which records a real tombstone and clears this entry)
    /// or the push is rejected (`restoreRow` clears it). Kept apart from
    /// the seq-keyed tombstones so a client fence never outranks a later
    /// legitimate re-create of the id. Mirrors the TS `optimisticTombstones`.
    private var optimisticTombstones: [String: Set<String>] = [:]

    private var listeners: [UUID: @Sendable () -> Void] = [:]

    /// Async persistence callback. Called from `applyChanges` /
    /// `applyChangesAsync` after merging into memory. Use to write through
    /// to SQLite or another durable store.
    public var persistFn: (@Sendable (ChangeEvent) async -> Void)?

    public init() {}

    // MARK: - Reads

    public func list(_ entity: String) -> [Row] {
        lock.lock(); defer { lock.unlock() }
        guard let table = tables[entity] else { return [] }
        return Array(table.values)
    }

    public func get(_ entity: String, id: String) -> Row? {
        lock.lock(); defer { lock.unlock() }
        return tables[entity]?[id]
    }

    // MARK: - Apply

    /// Apply a single change event. Tombstone-aware: insert/update events
    /// older than a recorded delete are silently dropped.
    public func applyChange(_ change: ChangeEvent) {
        lock.lock(); defer { lock.unlock() }
        applyChangeLocked(change)
    }

    /// Apply many changes and notify once. Persistence runs fire-and-forget
    /// — prefer `applyChangesAsync(_:)` if you'll advance a sync cursor
    /// after this call.
    public func applyChanges(_ changes: [ChangeEvent]) {
        lock.lock()
        for change in changes { applyChangeLocked(change) }
        let snapshot = changes.map { hydrateFromMemoryLocked($0) }
        let listeners = Array(self.listeners.values)
        let persist = persistFn
        lock.unlock()

        for l in listeners { l() }
        if let persist {
            for change in snapshot {
                Task.detached { await persist(change) }
            }
        }
    }

    /// Apply many changes and `await` the persistence writes before
    /// returning. Use this whenever you'll save a cursor after — otherwise
    /// a crash between memory-apply and disk-write can persist a cursor
    /// ahead of the on-disk replica.
    public func applyChangesAsync(_ changes: [ChangeEvent]) async {
        // The lock is acquired and released in a synchronous helper so the
        // compiler doesn't flag an NSLock call across an await boundary.
        let (snapshot, listeners, persist) = withLock {
            for change in changes { applyChangeLocked(change) }
            let snap = changes.map { hydrateFromMemoryLocked($0) }
            return (snap, Array(self.listeners.values), persistFn)
        }
        for l in listeners { l() }
        guard let persist else { return }
        await withTaskGroup(of: Void.self) { group in
            for change in snapshot {
                group.addTask { await persist(change) }
            }
        }
    }

    private func withLock<T>(_ body: () -> T) -> T {
        lock.lock(); defer { lock.unlock() }
        return body()
    }

    private func applyChangeLocked(_ change: ChangeEvent) {
        if tables[change.entity] == nil {
            tables[change.entity] = [:]
        }
        if (change.kind == .insert || change.kind == .update),
           isTombstonedLocked(entity: change.entity, id: change.row_id, atSeq: change.seq) {
            return
        }
        switch change.kind {
        case .insert:
            guard let data = change.data else { return }
            // Spread first, then force authoritative id — mirrors the TS
            // safety against a buggy server event corrupting the primary key.
            var merged = data
            merged["id"] = .string(change.row_id)
            tables[change.entity]?[change.row_id] = merged
        case .update:
            guard let data = change.data else { return }
            var existing = tables[change.entity]?[change.row_id] ?? ["id": .string(change.row_id)]
            for (k, v) in data { existing[k] = v }
            existing["id"] = .string(change.row_id)
            tables[change.entity]?[change.row_id] = existing
        case .delete:
            tables[change.entity]?.removeValue(forKey: change.row_id)
            recordTombstoneLocked(entity: change.entity, id: change.row_id, seq: change.seq)
        }
    }

    private func hydrateFromMemoryLocked(_ change: ChangeEvent) -> ChangeEvent {
        if change.kind == .delete { return change }
        guard let merged = tables[change.entity]?[change.row_id] else { return change }
        var copy = change
        copy.data = merged
        return copy
    }

    private func isTombstonedLocked(entity: String, id: String, atSeq: Int64) -> Bool {
        if optimisticTombstones[entity]?.contains(id) == true { return true }
        guard let tombSeq = tombstones[entity]?[id] else { return false }
        return atSeq < tombSeq
    }

    private func recordTombstoneLocked(entity: String, id: String, seq: Int64) {
        // A server-issued tombstone supersedes the optimistic fence for
        // this id, so a later legitimate re-create is not blocked.
        optimisticTombstones[entity]?.remove(id)
        if tombstones[entity] == nil { tombstones[entity] = [:] }
        let existing = tombstones[entity]?[id]
        if existing == nil || seq > (existing ?? 0) {
            tombstones[entity]?[id] = seq
        }
    }

    // MARK: - Optimistic

    /// Apply an optimistic insert under a freshly minted Pylon id and
    /// return it. The engine threads the same id through the push so the
    /// canonical server row replaces the ghost in place.
    @discardableResult
    public func optimisticInsert(_ entity: String, _ data: Row) -> String {
        let id = PylonIds.generate()
        optimisticInsertWithId(entity, id: id, data)
        return id
    }

    /// Apply an optimistic insert under a caller-chosen id. Mirrors the TS
    /// `optimisticInsertWithId`.
    public func optimisticInsertWithId(_ entity: String, id: String, _ data: Row) {
        lock.lock()
        if tables[entity] == nil { tables[entity] = [:] }
        var copy = data
        copy["id"] = .string(id)
        tables[entity]?[id] = copy
        let listeners = Array(self.listeners.values)
        lock.unlock()
        for l in listeners { l() }
    }

    public func optimisticUpdate(_ entity: String, id: String, _ data: Row) {
        lock.lock()
        guard var existing = tables[entity]?[id] else {
            lock.unlock(); return
        }
        for (k, v) in data { existing[k] = v }
        tables[entity]?[id] = existing
        let listeners = Array(self.listeners.values)
        lock.unlock()
        for l in listeners { l() }
    }

    /// Apply an optimistic delete: drop the row and fence the id against
    /// any insert/update until the server's own delete event arrives (or
    /// the push is rejected and `restoreRow` lifts the fence). The fence is
    /// not a seq tombstone: a server re-create of the id after the delete
    /// still lands once the server's delete has cleared it.
    public func optimisticDelete(_ entity: String, id: String) {
        lock.lock()
        tables[entity]?.removeValue(forKey: id)
        if optimisticTombstones[entity] == nil { optimisticTombstones[entity] = [] }
        optimisticTombstones[entity]?.insert(id)
        let listeners = Array(self.listeners.values)
        let persist = persistFn
        lock.unlock()
        for l in listeners { l() }
        // Mirror applyChange's persistence path so the delete survives
        // an app restart. Without this, the row sits in memory as
        // deleted but the on-disk replica still has it — relaunching
        // restores LocalStore from disk and the row reappears. The
        // pending mutation in the queue re-pushes the delete on launch.
        if let persist {
            let change = ChangeEvent(
                seq: 0,
                entity: entity,
                row_id: id,
                kind: .delete,
                data: nil,
                timestamp: ""
            )
            Task.detached { await persist(change) }
        }
    }

    // MARK: - Listeners

    @discardableResult
    public func subscribe(_ listener: @escaping @Sendable () -> Void) -> () -> Void {
        let id = UUID()
        lock.lock()
        listeners[id] = listener
        lock.unlock()
        return { [weak self] in
            self?.lock.lock()
            self?.listeners.removeValue(forKey: id)
            self?.lock.unlock()
        }
    }

    /// Manually fire all listeners. Used by `SyncEngine` after side-channel
    /// updates (presence, resolved-session changes) so subscribers re-read.
    public func notify() {
        lock.lock()
        let listeners = Array(self.listeners.values)
        lock.unlock()
        for l in listeners { l() }
    }

    /// Drop every table + tombstone in-place, then notify. Used on identity
    /// flip (token or tenant changed).
    public func clearAll() {
        lock.lock()
        tables.removeAll()
        tombstones.removeAll()
        optimisticTombstones.removeAll()
        let listeners = Array(self.listeners.values)
        lock.unlock()
        for l in listeners { l() }
    }

    // MARK: - Reconcile + rollback (TS parity)
    //
    // These mirror `packages/sync/src/local-store.ts`. The SyncEngine uses
    // them to sweep phantom rows (rows the server no longer returns) and to
    // undo optimistic mutations the server permanently rejected — the two
    // behaviors the Swift engine was missing, which surfaced as the Mac app
    // showing stale/ghost recordings that never cleared.

    /// Entity names that currently have a local table. Reconcile unions this
    /// with the engine's observed-entity set so a server row in a
    /// never-cached entity is still swept in.
    public func entityNames() -> [String] {
        lock.lock(); defer { lock.unlock() }
        return Array(tables.keys)
    }

    /// Reconcile a freshly-fetched server snapshot for one entity: upsert the
    /// `upserts`, and remove+tombstone every id in `removalIds`. The engine
    /// has already excluded rows with pending/failed mutations from both
    /// lists. `tombstoneSeq` is the engine's cursor at fetch time, so a later
    /// re-creation (higher seq) bypasses the tombstone and re-adds still flow.
    public func applyReconcileBatch(
        _ entity: String,
        upserts: [Row],
        removalIds: [String],
        tombstoneSeq: Int64
    ) {
        let (snapshot, listeners, persist) = withLock {
            () -> ([ChangeEvent], [@Sendable () -> Void], (@Sendable (ChangeEvent) async -> Void)?) in
            if tables[entity] == nil { tables[entity] = [:] }
            var snaps: [ChangeEvent] = []
            for row in upserts {
                guard case let .string(id)? = row["id"], !id.isEmpty else { continue }
                // A row tombstoned at a HIGHER seq is fresher — the upsert is
                // a stale pre-delete view; skip it.
                if isTombstonedLocked(entity: entity, id: id, atSeq: tombstoneSeq + 1) { continue }
                var merged = row
                merged["id"] = .string(id)
                tables[entity]?[id] = merged
                snaps.append(ChangeEvent(seq: tombstoneSeq, entity: entity, row_id: id, kind: .insert, data: merged, timestamp: ""))
            }
            for id in removalIds {
                if tables[entity]?.removeValue(forKey: id) != nil {
                    recordTombstoneLocked(entity: entity, id: id, seq: tombstoneSeq)
                    snaps.append(ChangeEvent(seq: tombstoneSeq, entity: entity, row_id: id, kind: .delete, data: nil, timestamp: ""))
                }
            }
            return (snaps, Array(self.listeners.values), persistFn)
        }
        guard !snapshot.isEmpty else { return }
        for l in listeners { l() }
        if let persist {
            for change in snapshot { Task.detached { await persist(change) } }
        }
    }

    /// Per-subscriber row revocation: the server signaled that this client
    /// lost read access to one row. ALWAYS records the tombstone (even when
    /// the row is not in memory) so a stale insert/update that arrives after
    /// the revocation cannot resurrect it. Returns whether a row was removed.
    /// Mirrors the TS `revokeRow`.
    public func revokeRow(_ entity: String, id: String, tombstoneSeq: Int64) -> Bool {
        lock.lock()
        let removed = tables[entity]?.removeValue(forKey: id) != nil
        recordTombstoneLocked(entity: entity, id: id, seq: tombstoneSeq)
        let listeners = removed ? Array(self.listeners.values) : []
        lock.unlock()
        for l in listeners { l() }
        return removed
    }

    /// Remove an optimistic insert ghost whose push was permanently rejected.
    /// No tombstone — the id was never real server state.
    public func rollbackOptimisticInsert(_ entity: String, id: String) {
        lock.lock()
        let removed = tables[entity]?.removeValue(forKey: id) != nil
        let listeners = removed ? Array(self.listeners.values) : []
        lock.unlock()
        for l in listeners { l() }
    }

    /// Restore a row to its captured pre-mutation value after a PERMANENTLY-
    /// rejected update/delete, clearing the optimistic-delete fence so the
    /// row (and future server events for it) aren't blocked. Only called when
    /// the server rejected the op (4xx) — so the row is unchanged server-side
    /// and the restore is authoritative. `prev == nil` removes the row (the
    /// pre-mutation state was "absent"). Persists the restored value so it
    /// survives the optimistic delete's on-disk tombstone.
    public func restoreRow(_ entity: String, id: String, prev: Row?) {
        let (snapshot, listeners, persist) = withLock {
            () -> (ChangeEvent, [@Sendable () -> Void], (@Sendable (ChangeEvent) async -> Void)?) in
            // The failed mutation's own optimistic fence always clears.
            optimisticTombstones[entity]?.remove(id)
            let snap: ChangeEvent
            if tombstones[entity]?[id] != nil {
                // The server removed this row while the rejected push was in
                // flight; its delete outranks the local rollback.
                tables[entity]?.removeValue(forKey: id)
                snap = ChangeEvent(seq: 0, entity: entity, row_id: id, kind: .delete, data: nil, timestamp: "")
            } else if let prev {
                if tables[entity] == nil { tables[entity] = [:] }
                var row = prev
                row["id"] = .string(id)
                tables[entity]?[id] = row
                snap = ChangeEvent(seq: 0, entity: entity, row_id: id, kind: .insert, data: row, timestamp: "")
            } else {
                tables[entity]?.removeValue(forKey: id)
                snap = ChangeEvent(seq: 0, entity: entity, row_id: id, kind: .delete, data: nil, timestamp: "")
            }
            return (snap, Array(self.listeners.values), persistFn)
        }
        for l in listeners { l() }
        if let persist { Task.detached { await persist(snapshot) } }
    }
}

File: packages/sync/src/local-store.ts
// ---------------------------------------------------------------------------
// In-memory replica of server state.
//
// Tracks tables (entity → id → row), tombstones (delete seqs that
// fence out late-arriving stale inserts), and optimistic tombstones
// (pending client deletes that block incoming inserts until the
// server's authoritative delete arrives). Persistence backends
// (IndexedDB) wire through `_persistFn`.
// ---------------------------------------------------------------------------

import type { ChangeEvent, Row } from "./types";

export class LocalStore {
  private tables: Map<string, Map<string, Row>> = new Map();
  /**
   * `(entity, row_id) → deletedAt seq`. A row in this map has been
   * deleted; any insert/update event older than its tombstone seq is
   * ignored so an out-of-order replay can't resurrect it. Without
   * this, a delete followed by a reconnect-driven replay of the
   * original insert would re-materialize the row.
   *
   * Real server-issued tombstones use the event's own seq. Optimistic
   * client tombstones live in `optimisticTombstones` instead.
   */
  private tombstones: Map<string, Map<string, number>> = new Map();
  /**
   * Pending optimistic deletes — `(entity, row_id)` pairs the client
   * has dropped but the server hasn't yet confirmed. Stored
   * separately from `tombstones` so the real server-issued delete
   * (with its real, smaller seq) can supersede the optimistic entry
   * without being max-merged out.
   *
   * Invariant: a future legitimate re-create of the same id must
   * succeed once the server's authoritative delete arrives. Test:
   * `optimistic_delete_releases_id_when_server_confirms`.
   */
  private optimisticTombstones: Map<string, Set<string>> = new Map();
  private listeners: Set<() => void> = new Set();

  /** Get all rows for an entity. */
  list(entity: string): Row[] {
    const table = this.tables.get(entity);
    if (!table) return [];
    return Array.from(table.values());
  }

  /** Get a row by ID. */
  get(entity: string, id: string): Row | null {
    return this.tables.get(entity)?.get(id) ?? null;
  }

  /** Snapshot of every entity name with at least one local row. Used
   *  by `SyncEngine.reconcile` to know which tables to diff against
   *  server truth. */
  entityNames(): string[] {
    const names: string[] = [];
    for (const [name, table] of this.tables) {
      if (table.size > 0) names.push(name);
    }
    return names;
  }

  /**
   * Remove a row whose absence was confirmed by the server-truth
   * reconciler. Records a tombstone at `tombstoneSeq` so a stale
   * insert/update replayed afterwards (e.g. a slow WS frame) can't
   * resurrect it. Callers pass the current sync cursor — any future
   * change events have higher seqs and pass the tombstone check,
   * while older replays are filtered.
   *
   * Differs from `optimisticDelete`: this is "the server says this
   * row is gone right now"; that one is "the client wants this row
   * gone before the server has confirmed."
   */
  reconcileRemove(entity: string, id: string, tombstoneSeq: number): boolean {
    const table = this.tables.get(entity);
    if (!table || !table.has(id)) return false;
    table.delete(id);
    this.recordTombstone(entity, id, tombstoneSeq);
    return true;
  }

  /**
   * Per-subscriber row revocation: the server signaled that the
   * current client lost read access to a specific row. ALWAYS
   * records the tombstone (even when the row isn't in memory), so
   * a stale insert/update event that arrives after the revocation
   * can't resurrect the row.
   *
   * `reconcileRemove` returns early when the row is absent — that's
   * the right behavior for "did anything change?" reconcile passes,
   * but wrong here: a CRDT-only consumer using `useLoroDoc` without
   * a JSON row never had the row in `tables` to begin with, and
   * without the tombstone any future server-issued insert
   * (legitimate re-grant or a slow stale frame) would land.
   *
   * Returns true if the row was present + removed; the tombstone
   * is recorded regardless.
   */
  revokeRow(entity: string, id: string, tombstoneSeq: number): boolean {
    const removed = this.tables.get(entity)?.delete(id) ?? false;
    this.recordTombstone(entity, id, tombstoneSeq);
    return removed;
  }

  private isTombstoned(entity: string, id: string, at_seq?: number): boolean {
    if (this.optimisticTombstones.get(entity)?.has(id)) return true;
    const tombSeq = this.tombstones.get(entity)?.get(id);
    if (tombSeq === undefined) return false;
    // No-seq caller means "this change has no provenance" — treat as
    // older than any tombstone (safer default than admitting it).
    if (at_seq === undefined) return true;
    return at_seq < tombSeq;
  }

  private recordTombstone(entity: string, id: string, seq: number): void {
    // A real (server-issued) tombstone supersedes any pending
    // optimistic entry for this id. Drop the optimistic guard so a
    // future legitimate re-create of the same id isn't blocked.
    this.optimisticTombstones.get(entity)?.delete(id);
    if (!this.tombstones.has(entity)) {
      this.tombstones.set(entity, new Map());
    }
    const existing = this.tombstones.get(entity)!.get(id);
    if (existing === undefined || seq > existing) {
      this.tombstones.get(entity)!.set(id, seq);
    }
  }

  /** Apply a change event to the local store. */
  applyChange(change: ChangeEvent): void {
    if (!this.tables.has(change.entity)) {
      this.tables.set(change.entity, new Map());
    }
    const table = this.tables.get(change.entity)!;

    // Drop insert/update events that arrive AFTER a delete for the
    // same row. The tombstone map records the seq of the delete;
    // anything strictly older is a stale resurrect.
    if (
      (change.kind === "insert" || change.kind === "update") &&
      this.isTombstoned(change.entity, change.row_id, change.seq)
    ) {
      return;
    }

    switch (change.kind) {
      case "insert":
        if (change.data) {
          // Spread data FIRST, then force `id = change.row_id` —
          // otherwise an `id` field in `data` could shadow the
          // authoritative row_id and corrupt the replica's primary key
          // on reload.
          table.set(change.row_id, {
            ...change.data,
            id: change.row_id,
          });
        }
        break;
      case "update":
        if (change.data) {
          const existing = table.get(change.row_id) ?? { id: change.row_id };
          table.set(change.row_id, {
            ...existing,
            ...change.data,
            id: change.row_id,
          });
        }
        break;
      case "delete":
        table.delete(change.row_id);
        this.recordTombstone(change.entity, change.row_id, change.seq);
        break;
    }
  }

  /** Apply multiple changes synchronously. Persistence runs fire-
   *  and-forget. Prefer `applyChangesAsync` when you plan to advance
   *  a cursor after — otherwise a crash can save the cursor before
   *  rows hit disk, causing permanent missed changes on restart. */
  applyChanges(changes: ChangeEvent[]): void {
    for (const change of changes) {
      this.applyChange(change);
    }
    this.notify();

    if (this._persistFn) {
      for (const change of changes) {
        const merged = this.hydrateFromMemory(change);
        void this._persistFn(merged);
      }
    }
  }

  /**
   * Apply + persist, awaiting disk writes before returning. Callers
   * that are about to advance a cursor based on `changes` MUST use
   * this path — otherwise cursor durability is broken: a crash
   * between the memory apply and the eventual disk write can persist
   * a cursor that's ahead of the replica, skipping those rows
   * forever on restart.
   *
   * Returns `true` when every persist write reached disk durably,
   * `false` when at least one degraded (quota / abort). The engine
   * uses the result to hold the PERSISTED cursor back: a row that
   * didn't reach disk must not be skipped by an advanced on-disk
   * cursor on the next cold start. The in-memory replica always
   * reflects the change regardless.
   */
  async applyChangesAsync(changes: ChangeEvent[]): Promise<boolean> {
    for (const change of changes) {
      this.applyChange(change);
    }
    this.notify();
    let allDurable = true;
    if (this._persistFn) {
      // Sequential await — concurrent IDB writes can resolve out of
      // order, racing an update behind its own delete on disk. The
      // engine's `applyQueue` sequences events into here; we
      // preserve that order down to the disk layer.
      for (const change of changes) {
        const result = this._persistFn(this.hydrateFromMemory(change));
        if (result instanceof Promise) {
          const durable = await result;
          if (durable === false) allDurable = false;
        }
      }
    }
    return allDurable;
  }

  /**
   * Reshape a change event so its `data` field matches the row as it
   * now exists in memory after `applyChange` merged the patch.
   * Persistence callers (IndexedDB) save the full row, which only
   * works if they receive the full row. Deletes pass through
   * untouched.
   */
  private hydrateFromMemory(change: ChangeEvent): ChangeEvent {
    if (change.kind === "delete") return change;
    const merged = this.tables.get(change.entity)?.get(change.row_id);
    if (!merged) return change;
    return { ...change, data: merged };
  }

  /** Persistence callback for auto-saving changes. Returns
   *  `Promise<boolean>` (true = durable, false = degraded) so
   *  `applyChangesAsync` can gate the on-disk cursor on durability.
   *  Void-returning callbacks are accepted for backwards compatibility
   *  (treated as durable / fire-and-forget). */
  _persistFn: ((change: ChangeEvent) => void | Promise<boolean>) | null = null;

  /** Subscribe to store changes. Returns unsubscribe function. */
  subscribe(listener: () => void): () => void {
    this.listeners.add(listener);
    return () => this.listeners.delete(listener);
  }

  notify(): void {
    for (const listener of this.listeners) {
      listener();
    }
  }

  /** Apply an optimistic insert. Returns a temporary id. */
  optimisticInsert(entity: string, data: Row): string {
    const tempId = `_pending_${Date.now()}_${Math.random().toString(36).slice(2)}`;
    if (!this.tables.has(entity)) {
      this.tables.set(entity, new Map());
    }
    this.tables.get(entity)!.set(tempId, { id: tempId, ...data });
    this.notify();
    return tempId;
  }

  /**
   * Apply an optimistic insert with a caller-provided id.
   *
   * Used by `useMutation({ optimistic })`: the React hook generates a
   * Pylon-shaped id (40-char hex via `generateId()`), threads it
   * through the mutation args as `_optimisticId`, and the server
   * function honors it on `ctx.db.insert("Entity", { id, ... })`.
   * Because the optimistic ghost and the canonical row share the
   * same row_id, the WS broadcast that follows lands as a field-
   * level merge on top of the optimistic — no delete-then-replace
   * flash, no temp-row swap.
   */
  optimisticInsertWithId(entity: string, id: string, data: Row): void {
    if (!this.tables.has(entity)) {
      this.tables.set(entity, new Map());
    }
    this.tables.get(entity)!.set(id, { ...data, id });
    this.notify();
  }

  /**
   * Roll back an optimistic insert without leaving a tombstone.
   *
   * Counterpart to `optimisticInsertWithId`. On rejection the ghost
   * row must go away, but we MUST NOT leave a tombstone — a future
   * legitimate insert with the same id (user retries, workflow
   * eventually creates the row) must not be blocked.
   * `optimisticDelete` records a MAX_SAFE_INTEGER tombstone, which is
   * the wrong semantic here; this is a plain remove.
   */
  rollbackOptimisticInsert(entity: string, id: string): void {
    const removed = this.tables.get(entity)?.delete(id);
    if (removed) this.notify();
  }

  /** Apply an optimistic update. */
  optimisticUpdate(entity: string, id: string, data: Partial<Row>): void {
    const table = this.tables.get(entity);
    if (!table) return;
    const existing = table.get(id);
    if (existing) {
      table.set(id, { ...existing, ...data });
      this.notify();
    }
  }

  /**
   * Undo a rejected optimistic update/delete by restoring the row to its
   * captured pre-mutation value and clearing any optimistic tombstone
   * for it. `failPushedMutation` calls this when the server rejects an
   * update (restore the prior field values) or a delete (bring the row
   * back AND un-fence it so the row — and any future server insert of
   * the id — isn't blocked by the lingering optimistic tombstone).
   *
   * `prev === null` means the row didn't exist before the mutation
   * (e.g. an update on a row that was itself an un-acked insert) — in
   * that case we just remove it + clear the fence.
   *
   * A REAL (server-issued) tombstone wins over the restore: if an
   * authoritative delete/revocation for this id landed on the applyQueue
   * while the rejected push was in flight (the opQueue and applyQueue run
   * independently), resurrecting `prev` here would briefly un-delete a row
   * the server says is gone — healed only at the next reconcile. So when a
   * server tombstone is present we drop the row and let the canonical
   * state stand; the failed mutation's own optimistic fence is cleared
   * regardless so a later legitimate re-create of the id isn't blocked.
   */
  restoreRow(entity: string, id: string, prev: Row | null): void {
    // The failed mutation's own optimistic fence always clears.
    this.optimisticTombstones.get(entity)?.delete(id);
    if (this.tombstones.get(entity)?.has(id)) {
      // Server authoritatively removed this row mid-flight — its
      // deletion outranks our local rollback.
      this.tables.get(entity)?.delete(id);
      this.notify();
      return;
    }
    if (prev) {
      if (!this.tables.has(entity)) this.tables.set(entity, new Map());
      this.tables.get(entity)!.set(id, prev);
    } else {
      this.tables.get(entity)?.delete(id);
    }
    this.notify();
  }

  /** Apply an optimistic delete. Block any incoming insert/update
   *  for this id until the server's authoritative delete arrives. */
  optimisticDelete(entity: string, id: string): void {
    this.tables.get(entity)?.delete(id);
    if (!this.optimisticTombstones.has(entity)) {
      this.optimisticTombstones.set(entity, new Set());
    }
    this.optimisticTombstones.get(entity)!.add(id);
    this.notify();
  }

  /**
   * Apply a reconcile result against the local replica. Reconcile is
   * "the server says these are the canonical rows right now" — it does
   * NOT carry per-event seqs. Distinct from `applyChange` (which
   * threads server-issued change events through a monotonic seq
   * guard + cursor advance).
   *
   * `tombstoneSeq` is the cursor at the start of the reconcile fetch.
   * - Upserts skip rows whose tombstone is fresher (would be stale
   *   resurrections of a later delete).
   * - Removals tombstone at `tombstoneSeq` so a later legitimate
   *   re-create (with a strictly greater seq) flows through.
   *
   * Cursor is NOT advanced — reconcile fabricates no real seqs, so
   * the engine's cursor stays pinned to the change-log position the
   * reconcile started from.
   */
  async applyReconcileBatch(
    entity: string,
    upserts: Row[],
    removalIds: string[],
    tombstoneSeq: number,
  ): Promise<void> {
    if (!this.tables.has(entity)) this.tables.set(entity, new Map());
    const table = this.tables.get(entity)!;
    const applied: Array<{ id: string; row: Row }> = [];
    for (const row of upserts) {
      const id = (row as { id?: unknown }).id;
      if (typeof id !== "string" || id.length === 0) continue;
      // Treat the upsert as effective at `tombstoneSeq + 1`. Anything
      // tombstoned at a HIGHER seq is fresher and wins; the upsert is
      // a stale view that pre-dated the delete.
      if (this.isTombstoned(entity, id, tombstoneSeq + 1)) continue;
      const merged = { ...row, id };
      table.set(id, merged);
      applied.push({ id, row: merged });
    }
    const removed: string[] = [];
    for (const id of removalIds) {
      if (this.reconcileRemove(entity, id, tombstoneSeq)) {
        removed.push(id);
      }
    }
    if (applied.length > 0 || removed.length > 0) this.notify();
    // Persist sequentially so disk order matches memory order — same
    // discipline as applyChangesAsync.
    if (this._persistFn) {
      for (const { id, row } of applied) {
        const ev: ChangeEvent = {
          seq: tombstoneSeq + 1,
          entity,
          row_id: id,
          kind: "insert",
          data: row,
          timestamp: "",
        };
        const result = this._persistFn(ev);
        if (result instanceof Promise) await result;
      }
      for (const id of removed) {
        const ev: ChangeEvent = {
          seq: tombstoneSeq,
          entity,
          row_id: id,
          kind: "delete",
          data: undefined as unknown as Row,
          timestamp: "",
        };
        const result = this._persistFn(ev);
        if (result instanceof Promise) await result;
      }
    }
  }

  /**
   * Drop every table + tombstone in-place, then notify. Used by the
   * sync engine's `resetReplica()` on identity flip (token or tenant
   * changed — the old replica reflects a different visible set).
   */
  clearAll(): void {
    this.tables.clear();
    this.tombstones.clear();
    this.optimisticTombstones.clear();
    this.notify();
  }
}
